### PR TITLE
[iterlib] Use Iterator<T> instead of Iterator

### PR DESCRIPTION
**Pull request by adsharma**
Someday I hope people can write code such as:

`
for (const auto& v : kvstore(shard, prefix)) {
  StronglyTyped t = v.schematize(Schema);
  ... 
}
`

This diff doesn't quite achieve that. But is a step in that direction. Problems to solve: we need an
iterator_adaptor that can do:

`it->prepare(); it++; // prepare and position on the first item`

Verified that sizeof(Iterator) and sizeof(AndIterator) didn't change as a result of this diff. The text size went up a bit from 3.3MB -> 3.8MB.

### CODE REVIEW

File: include/iterlib/OrIterator-inl.h [hunk 1]
```diff
@@ -188,9 +189,5 @@ bool ConcatIterator<T>::doNext() {
   return true;
 }
 
-// sorted-merge
```
Comment (yfeldblum) [inline include/iterlib/OrIterator-inl.h]: Make some actual template aliases instead of just comments?

Comment (adsharma) [inline include/iterlib/OrIterator-inl.h]: Aliases exist in OrIterator.h:

```
using UnionIterator = detail::UnionIterator<Item>;
using SortedMergeIterator = detail::SortedMergeIterator<Item>;
using ConcatIterator = detail::ConcatIterator<Item>;
```

**Comment by adsharma** [comment]
Thanks for reviewing. Added extern templates in the updated pull request.

**Comment by adsharma** [comment]
This seems to have broken the build on ubuntu-16.04/gcc because of what looks like the same issue here:

http://stackoverflow.com/questions/38929950/automatic-constructor-in-explicitly-instantiated-class-template

Error I get is:

```
undefined reference to `iterlib::detail::Iterator<iterlib::Item>::Iterator(iterlib::detail::Iterator<iterlib::Item>&&)'
```

**Comment by adsharma** [comment]
@yfeldblum : The explicit instantiation is already present in the C++ files (eg: AndIterator.cpp). Let me know if I missed any specific instances.

The linker error is specifically because of compiler behavior on methods that have '= default'.

**Comment by adsharma** [comment]
I reverted `a857c32 Add extern templates` until we have a solution for the linker problems.

**Pull request merged**